src/types.rs

```diff
@@ -1110,4 +1270,65 @@ mod tests {
 
         assert_eq!(new_sm.debug_id, Some(DebugId::default()));
     }
+
+    #[test]
+    fn test_compose_identity_left() {
```

Comment:
The tests are still very hard to understand, and they are only testing `identity` which is not a good test.

As you mention `magic-string` in the fn docs, you can use it as dev-dependency.
I would suggest to do a couple of tests such as this:

- Start with an empty file
- Use magic-string to insert a bunch of tokens into it
- Take a snapshot of that as a `&str` + Sourcemap as "original"
- Use that `&str` as a new base for another magic-string editing session
- Do a bunch of modifications on *both* editing sessions like: insert more snippets, delete parts of the code, move things around
- Output the delta as the "adjustment" sourcemap
- The "original" + "adjustment" composition should now be equivalent to snapshotting the original editing session once again

I hope this makes sense. You can even fuzz/proptest this with random modifications, as they should yield the same end result ideally.

Reply:
It makes sense, but the property in your last bullet point doesn't hold, the way the algorithm works now. If you
1. start with a string
2. do nothing, recording a sourcemap `ident.map`
3. make some modifications, recording a sourcemap `mod.map`

then `adjust_mappings(ident.map, mod.map)` should be the same as `mod.map`, but it's not. That can be achieved, but it's a little more complicated.